Задачи

dmitrirazumov opened the pull request:
//

Commit:
Решена задача №1

Commit:
Решена последняя тривиальная

Commit:
Решена простая №1

Commit:
Решена последняя простая

Commit:
programm is great!

dmitrirazumov committed:
easy programm (2)

dmitrirazumov committed:
easy programm (3)

dmitrirazumov committed:
easy programm (4)

dmitrirazumov committed:
programm normal level

dmitrirazumov committed:
programm normal level

dmitrirazumov committed:
circleInside

dmitrirazumov committed:
lcm (lesson 3)

dmitrirazumov committed:
revert (lesson 3)

dmitrirazumov committed:
2 programms from 4 lesson

dmitrirazumov committed:
5 programms from 4 lesson

dmitrirazumov committed:
5 programms from 4 lesson

dmitrirazumov committed:
5 programms from 4 lesson

dmitrirazumov committed:
1 programms from 4 lesson

mglukhikh commented:
Старайтесь в будущем не забывать об этом, пожалуйста. Просматривать все 94 репозитория на предмет изменений у нас нет возможности, так что, пока вы на кнопку не нажали, мы вынуждены считать, что вы ничего не сделали.

oddishwolf commented:
Сейчас все загрузил, есть шанс на исправление оценки? 


>Понедельник, 31 октября 2016, 12:54 +03:00 от Mikhail Glukhikh <notifications@github.com>:
>
>Старайтесь в будущем не забывать об этом, пожалуйста. Просматривать все 94 репозитория на предмет изменений у нас нет возможности, так что, пока вы на кнопку не нажали, мы вынуждены считать, что вы ничего не сделали.
>—
>You are receiving this because you authored the thread.
>Reply to this email directly,  view it on GitHub , or  mute the thread .

mglukhikh commented:
> Сейчас все загрузил, есть шанс на исправление оценки? 

Сейчас сообщать куратору, что вы злостный двоечник, я не буду. Но в конце семестра повторение подобной ситуации вполне может привести, например, к неполучению автомата или к снижению заслуженной оценки на балл.

dmitrirazumov committed:
1 programms from 4 lesson

dmitrirazumov committed:
2 programms from 5 lesson

mglukhikh commented:
recheck all

mglukhikh commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -44,7 +49,15 @@ fun ageDescription(age: Int): String = TODO()
+    val h = (s1 + s2 + s3)/2
+    return if (h <= s1) h/v1
+     else if ((h > s1) and (h <= s1 +s2))  t1+((h-s1)/v2)
+             else t1 + t2 + ((h - s1 - s2)/v3)
```
1) Старайтесь расставлять пробелы консистентно (то есть всюду одинаково). А то у вас в одном месте `h/v1`, в другом `s1 +s2`, в третьем `v1 * t1`. Это можно исправить автоматически с помощью авто-форматирования кода: Ctrl+Alt+L.

mglukhikh replied:
2) В этом месте лучше использовать конструкцию `when`, код будет выглядеть аккуратнее

mglukhikh commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -56,7 +69,12 @@ fun timeForHalfWay(t1: Double, v1: Double,
+    return if (((kingX != rookX1) and (kingY != rookY1)) and ((kingX != rookX2) and (kingY != rookY2))) 0
+    else if (((kingX == rookX1) or (kingY == rookY1)) and ((kingX != rookX2) and (kingY != rookY2))) 1
+    else if (((kingX != rookX1) and (kingY != rookY1)) and ((kingX == rookX2) or (kingY == rookY2))) 2
+    else 3
```
Здесь и далее, лучше использовать `&&` вместо `and`, `||` вместо `or`.

mglukhikh commented on src/lesson2/task2/Logical.kt:
```diff
@@ -17,7 +17,13 @@ fun pointInsideCircle(x: Double, y: Double, x0: Double, y0: Double, r: Double) =
-fun isNumberHappy(number: Int): Boolean = TODO()
+fun isNumberHappy(number: Int): Boolean {
+    val n1 = ((number%10) + ((number/10)%10))
+    val n2 = ((number/1000) + ((number/100)%10))
```
Пробелы

mglukhikh commented on src/lesson2/task2/Logical.kt:
```diff
@@ -17,7 +17,13 @@ fun pointInsideCircle(x: Double, y: Double, x0: Double, y0: Double, r: Double) =
+    val n1 = ((number%10) + ((number/10)%10))
+    val n2 = ((number/1000) + ((number/100)%10))
+    if (n1 == n2) return true
+     else return false
```
`if (x) return true else return false` эквивалентно `return x`.

mglukhikh commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -79,7 +97,31 @@ fun rookOrBishopThreatens(kingX: Int, kingY: Int,
+        if ((a == b) and (a > c)) return 0
+        if ((b > a) and (b == c)) return 0
+        return -1
+    } else return -1
```
Подумайте, нельзя ли решить эту задачу короче

mglukhikh commented on src/lesson2/task2/Logical.kt:
```diff
@@ -35,7 +41,11 @@ fun queenThreatens(x1: Int, y1: Int, x2: Int, y2: Int): Boolean = TODO()
+                 x2: Double, y2: Double, r2: Double): Boolean {
+    val length = Math.sqrt(sqr(x2-x1) + sqr(y2-y1))
+    if (length <= (r2-r1)) return true
+    else return false
```
1) Пробелы 
2) `if (x) return true else return false`

mglukhikh commented on src/lesson3/task1/Loop.kt:
```diff
@@ -73,7 +84,19 @@ fun fib(n: Int): Int = TODO()
+    val count = num_m + num_n
+    return multiplication/count
+
+
```
Пробелы, ненужные пустые строки

mglukhikh commented on src/lesson3/task1/Loop.kt:
```diff
@@ -131,7 +154,15 @@ fun cos(x: Double, eps: Double): Double = TODO()
+    var m = 0
+    var num = n
+    do {
+        m = (m*10 + (num%10))
```
Пробелы

mglukhikh commented on src/lesson3/task1/Loop.kt:
```diff
@@ -57,7 +57,18 @@ fun digitCountInNumber(n: Int, m: Int): Int =
+        }
+        return count
+    }
+    else return 0 // НЕ ДОДЕЛАНА !!!
```
Функция будет работать неверно с отрицательными `n`

mglukhikh commented on src/lesson4/task1/List.kt:
```diff
@@ -104,22 +105,45 @@ fun buildSumExample(list: List<Int>) = list.joinToString(separator = " + ", post
  */
-fun abs(v: List<Double>): Double = TODO()
+fun abs(v: List<Double>): Double {
+    var abswithoutsqrt = 0.0
```
camelCase: `absWithoutSqrt`

mglukhikh commented on src/lesson4/task1/List.kt:
```diff
@@ -104,22 +105,45 @@ fun buildSumExample(list: List<Int>) = list.joinToString(separator = " + ", post
-fun abs(v: List<Double>): Double = TODO()
+fun abs(v: List<Double>): Double {
+    var abswithoutsqrt = 0.0
+    for (i in 0..(v.size - 1)) {
```
Используйте цикл по элементам: `for (elem in v) { ... }` (см. пример negativeList)

mglukhikh commented on src/lesson4/task1/List.kt:
```diff
@@ -104,22 +105,45 @@ fun buildSumExample(list: List<Int>) = list.joinToString(separator = " + ", post
+        val sum = list.sum()
+    if (k == 0) return 0.0
+    if (k == 1) return sum
+    else return sum/k
```
В начале функции поехало форматирование. Пробелы в `sum/k`

mglukhikh commented on src/lesson4/task1/List.kt:
```diff
@@ -104,22 +105,45 @@ fun buildSumExample(list: List<Int>) = list.joinToString(separator = " + ", post
  */
-fun center(list: MutableList<Double>): MutableList<Double> = TODO()
+fun center(list: MutableList<Double>): MutableList<Double> {
+    if (list.isNotEmpty()) {
```
Эта проверка не нужна

mglukhikh replied:
Эта проверка по-прежнему не нужна

mglukhikh commented on src/lesson4/task1/List.kt:
```diff
@@ -148,7 +182,18 @@ fun polynom(p: List<Double>, x: Double): Double = TODO()
+        val large = list.size-1
+        for (i in 0..large) {
+            sum += list[i]
+            list[i]=sum
```
Пробелы

mglukhikh commented on src/lesson4/task1/List.kt:
```diff
@@ -157,15 +202,47 @@ fun accumulate(list: MutableList<Double>): MutableList<Double> = TODO()
+            number /= i
+        } else i += 2
+    }
+    result.sorted()
```
Этот вызов функции не нужен

mglukhikh replied:
Этот вызов функции по-прежнему не нужен

mglukhikh commented on src/lesson4/task1/List.kt:
```diff
@@ -157,15 +202,47 @@ fun accumulate(list: MutableList<Double>): MutableList<Double> = TODO()
+            result.add(i)
+            number /= i
+        } else i += 2
+    }
```
В этой функции не нужно повторять код `factorize`, её следует просто вызвать

mglukhikh replied:
Замечание верное, но не на том месте

mglukhikh replied:
Исправлено

mglukhikh commented on src/lesson5/task1/Parse.kt:
```diff
@@ -60,7 +60,30 @@ fun main(args: Array<String>) {
+            "ноября" -> 11
+            "декабря" -> 12
+            else -> 0
+        }
```
Для преобразования названия месяца в число создайте список названий (лучше на верхнем уровне, то есть перед функцией) и используйте `list.indexOf`.

mglukhikh replied:
Повторно! Для преобразования названия месяца в число создайте список названий (лучше на верхнем уровне, то есть перед функцией) и используйте `list.indexOf`.

mglukhikh commented on src/lesson5/task1/Parse.kt:
```diff
@@ -69,7 +92,30 @@ fun dateStrToDigit(str: String): String = TODO()
+            "11" -> "ноября"
+            "12" -> "декабря"
+            else -> 0
+        }
```
Здесь тоже можно использовать список названий месяцев (просто обратиться к нему по индексу).

mglukhikh replied:
Повторно! Здесь тоже можно использовать список названий месяцев (просто обратиться к нему по индексу).

mglukhikh reviewed (changes requested):
Пара задач решена с ошибкой (см. отчёт бота). Плюс замечания по стилю. Пожалуйста, исправьте замечания и ошибки, прежде чем решать другие задачию

dmitrirazumov committed:
1 programm from 6 lesson

mglukhikh commented:
Пожалуйста, исправьте замечания и ошибки в уроках 1-5, прежде чем решать задачи из следующих уроков. Задачи 6-го урока я проверю позже

dmitrirazumov committed:
normal and hard programms from 5 lesson

mglukhikh commented:
recheck all

mglukhikh commented:
Замечания из последнего Review по-прежнему в силе. Также, две задачи по-прежнему решены неверно

dmitrirazumov committed:
hard programm from 5 lesson

dmitrirazumov committed:
Добавил дополнительное условие в программу

dmitrirazumov committed:
Добавил дополнительное условие в программу

dmitrirazumov committed:
Добавил дополнительное условие в программу

dmitrirazumov committed:
Добавил дополнительное условие в программу

dmitrirazumov committed:
Добавил дополнительное условие в программу

dmitrirazumov committed:
Добавил дополнительное условие в программу

dmitrirazumov committed:
Добавил дополнительное условие в программу

dmitrirazumov committed:
Добавил дополнительное условие в программу

dmitrirazumov committed:
Добавил дополнительное условие в программу

dmitrirazumov committed:
Добавил дополнительное условие в программу

dmitrirazumov committed:
Добавил дополнительное условие в программу

dmitrirazumov commented:
Можете объяснить что не так с данной задачей? Уже перепробовал массу вариантов, но бот выводит ошибку. Я практически уверен, что ответ должен быть правильным:
/**
- Сложная
  *
- В строке представлено выражение вида "2 + 31 - 40 + 13",
- использующее целые положительные числа, плюсы и минусы, разделённые пробелами.
- Наличие двух знаков подряд "13 + + 10" или двух чисел подряд "1 2" не допускается.
- Вернуть значение выражения (6 для примера).
- Про нарушении формата входной строки бросить исключение IllegalArgumentException
  */
  fun plusMinus(expression: String): Int {
  val members = expression.split(" ")
  var result = members[0].toInt()
  val LengthOfString = members.size
  val RightNumber = Regex("""\d""")
  val RightSymbol = Regex("""[+-]""")
  val NotRightFormat = Regex("""[^\d+-]""")
  if ((LengthOfString == 1) && (members[0].contains(NotRightFormat)) || expression == "") throw IllegalArgumentException("Wrong")
  else {
      for (i in 0..LengthOfString - 2) {
          val TwoPluses = members[i].contains(RightSymbol) && members[i + 1].contains(RightSymbol)
          val TwoNumbers = members[i].contains(RightNumber) && members[i + 1].contains(RightNumber)
          val AllNotRightSymbols = members[i].contains(NotRightFormat) || members[i + 1].contains(NotRightFormat)
          val WrongString = TwoPluses || TwoNumbers || AllNotRightSymbols
          if (WrongString) throw IllegalArgumentException("Wrong")
          else {
              if (members[i] == "+") result += members[i + 1].toInt()
              if (members[i] == "-") result -= members[i + 1].toInt()
          }
      }
  }
  return result
  }
   

dmitrirazumov committed:
Добавил дополнительное условие в программу

dmitrirazumov committed:
Добавил дополнительное условие в программу

dmitrirazumov committed:
Добавил дополнительное условие в программу

dmitrirazumov committed:
Добавил дополнительное условие в программу

dmitrirazumov committed:
Добавил дополнительное условие в программу

dmitrirazumov committed:
2 programms from 4 lesson

dmitrirazumov committed:
Проверка правильности программы через бота

dmitrirazumov committed:
Проверка правильности программы через бота

dmitrirazumov committed:
Проверка правильности программы через бота

dmitrirazumov committed:
Проверка правильности программы через бота

dmitrirazumov committed:
Проверка правильности программы через бота

dmitrirazumov committed:
Проверка правильности программы через бота

dmitrirazumov committed:
Проверка правильности программы через бота

dmitrirazumov committed:
Проверка правильности программы через бота

dmitrirazumov committed:
Выяснение причины "Failure: Check Ended"

dmitrirazumov commented:
Здравствуйте. Я там кажется "немного" перестарался и получил вот это. Если знаете как с этим бороться и что делать, напишите пожалуйста. Я не могу теперь пушить задачи и проверять их через бота, так как постоянно пишет "Failure: Check Ended". Иными словами, сейчас не в состоянии загрузить решенную задачу для вашей проверки. (фотографии прилагаются на майле ну и, собственно, вы можете сами все видеть по последним коммитам, все перечеркнуто) 

Total: 22 / 33

Example: 9 / 9
Easy: 2 / 2
Normal: 8 / 10
Hard: 3 / 11
Impossible: 0 / 1

mglukhikh commented:
recheck all

mglukhikh commented:
Вроде бы тестирование успешно прошло. Возможно, это был временный эффект, связанный с проблемами сервера, на котором работает бот.

mglukhikh commented on src/lesson2/task2/Logical.kt:
```diff
@@ -17,7 +18,13 @@ fun pointInsideCircle(x: Double, y: Double, x0: Double, y0: Double, r: Double) =
+    val n1 = ((number % 10) + ((number / 10) % 10))
+    val n2 = ((number / 1000) + ((number / 100) % 10))
+    if (n1 == n2) return true
+    else return false
```
`if (x) return true else return false` эквивалентно `return x`

mglukhikh commented on src/lesson2/task2/Logical.kt:
```diff
@@ -35,7 +42,11 @@ fun queenThreatens(x1: Int, y1: Int, x2: Int, y2: Int): Boolean = TODO()
+                 x2: Double, y2: Double, r2: Double): Boolean {
+    val length = Math.sqrt(sqr(x2 - x1) + sqr(y2 - y1))
+    if (length <= (r2 - r1)) return true
+    else return false
```
См. выше

mglukhikh commented on src/lesson3/task1/Loop.kt:
```diff
@@ -73,7 +85,19 @@ fun fib(n: Int): Int = TODO()
+    var num_m = m
+    var num_n = n
+    val multiplication = (num_m * num_n)
+    while ((num_m != 0) and (num_n != 0)) {
```
Лучше использовать ленивый `&&`

mglukhikh commented on src/lesson4/task1/List.kt:
```diff
@@ -157,15 +201,47 @@ fun accumulate(list: MutableList<Double>): MutableList<Double> = TODO()
+            number /= i
+        } else i += 2
+    }
+    result.sorted()
```
Вы здесь дублируете код `factorize`, её нужно просто вызвать

mglukhikh commented on src/lesson4/task1/List.kt:
```diff
@@ -174,7 +250,20 @@ fun factorizeToString(n: Int): String = TODO()
+    val result = mutableListOf<Int>()
+    if (n < base) result.add(n)
+    else {
+        var Num = n
```
С маленькой буквы

mglukhikh commented on src/lesson5/task1/Parse.kt:
```diff
@@ -95,7 +140,19 @@ fun flattenPhoneNumber(phone: String): String = TODO()
+fun bestLongJump(jumps: String): Int {
+    val members = jumps.split(" ")
+    val NotTheRightMembers = Regex("""[^\d%-]""")
+    val RightNumber = Regex("""\d""")
```
Переменные должны называться с маленькой буквы

mglukhikh commented on src/lesson5/task1/Parse.kt:
```diff
@@ -95,7 +140,19 @@ fun flattenPhoneNumber(phone: String): String = TODO()
+    val NotTheRightMembers = Regex("""[^\d%-]""")
+    val RightNumber = Regex("""\d""")
+    var max = -1
+    for (i in 0..members.size - 1) {
```
Здесь лучше применить цикл по элементам: `for (member in members) { ... }`

mglukhikh commented on src/lesson5/task1/Parse.kt:
```diff
@@ -118,7 +183,29 @@ fun bestHighJump(jumps: String): Int = TODO()
+    val LengthOfString = members.size
+    val RightNumber = Regex("""\d""")
+    val RightSymbol = Regex("""[+-]""")
+    val NotRightFormat = Regex("""[^\d+-]""")
```
Переменные следует называть с маленькой буквы

mglukhikh commented on src/lesson5/task1/Parse.kt:
```diff
@@ -118,7 +183,29 @@ fun bestHighJump(jumps: String): Int = TODO()
-fun plusMinus(expression: String): Int = TODO()
+fun plusMinus(expression: String): Int {
+    val members = expression.split(" ")
+    var result = members[0].toInt()
```
Что будет, если `members[0]` окажется не числом? Например, для `expression = "BAD"`.

dmitrirazumov replied:
Так это у меня и есть основная проверка

dmitrirazumov replied:
val notRightFormat = Regex("""[^\d+-]""")
if ((lengthOfString == 1) && **_((members[0].contains(notRightFormat))_** || expression == "")) throw IllegalArgumentException(" ") #

mglukhikh replied:
Это замечательно, но вы до этой строчки просто не дойдёте. Так как словите исключение здесь.

mglukhikh reviewed (changes requested):
Пересмотрел уроки 2-5. Дал вам подсказку по поводу `plusMinus`. Осталось в силе много старых замечаний, так что к уроку 6 пока приступать рано.

dmitrirazumov committed:
Исправление замечаний

oddishwolf commented on src/lesson5/task1/Parse.kt:
```diff
@@ -118,7 +183,29 @@ fun bestHighJump(jumps: String): Int = TODO()
-fun plusMinus(expression: String): Int = TODO()
+fun plusMinus(expression: String): Int {
+    val members = expression.split(" ")
+    var result = members[0].toInt()
```
val notRightFormat = Regex("""[^\d+-]""")
...
if ((lengthOfString == 1) &&#  ((members[0].contains(notRightFormat)) || expression == "")) throw IllegalArgumentException(" ")

dmitrirazumov committed:
Пытка программы "PlusMinus"

mglukhikh commented on src/lesson3/task1/Loop.kt:
```diff
@@ -57,7 +58,18 @@ fun digitCountInNumber(n: Int, m: Int): Int =
+            number /= 10
+            k += 1
+        }
+    }
```
Может исправите эту задачу наконец? Она работает неверно для отрицательных чисел

mglukhikh commented on src/lesson4/task1/List.kt:
```diff
@@ -128,7 +150,16 @@ fun center(list: MutableList<Double>): MutableList<Double> = TODO()
  */
-fun times(a: List<Double>, b: List<Double>): Double = TODO()
+fun times(a: List<Double>, b: List<Double>): Double {
+    var C = 0.0
```
С маленькой буквы

mglukhikh reviewed (changes requested):
Замечаний по коду почти не осталось. Не работают `digitNumber` и `plusMinus`. Исправьте их, и я закрою этот Pull Request.

dmitrirazumov committed:
Пытка программы "PlusMinus"

dmitrirazumov committed:
Исправление всех ошибок

dmitrirazumov commented:
Исправил все ошибки. Программы прошли проверку бота.

mglukhikh commented:
recheck all

mglukhikh commented:
Задача `mostExpensive` у вас не проходит тесты (причём даже локальные в учебном проекте). Доразбирайтесь уж с ней.

dmitrirazumov committed:
Most Expensive + one program from 6 lesson (task 1)

dmitrirazumov committed:
//

mglukhikh commented on src/lesson5/task1/Parse.kt:
```diff
@@ -187,4 +273,34 @@ fun fromRoman(roman: String): Int = TODO()
+/**
+ *
+ */
+fun myFun(from: String, to: String, route: String): Int {
```
Пожалуйста, удалите эту функцию из проекта

mglukhikh commented on src/lesson5/task1/Parse.kt:
```diff
@@ -118,7 +161,30 @@ fun bestHighJump(jumps: String): Int = TODO()
+            val TwoPluses = members[i].contains(rightSymbol) && members[i + 1].contains(rightSymbol)
+            val TwoNumbers = members[i].contains(rightNumber) && members[i + 1].contains(rightNumber)
+            val AllNotRightSymbols = members[i].contains(notRightFormat) || members[i + 1].contains(notRightFormat)
+            val WrongString = TwoPluses || TwoNumbers || AllNotRightSymbols
```
Напоминаю вам, что переменные следует называть в `camelCase`, например, `wrongString` (см. также ещё три переменные выше)

mglukhikh reviewed (changes requested):
В целом всё в порядке, два незначительных замечания. Вы можете, по своему желанию, решать следующие задачи 5-го или 6-го урока или же перейти к 7-му. Судя по результатам тестирования, потратить ещё немножко времени на задачи уроков 5 и 6 было бы для вас полезно. Отмечу, например, что задача `plusMinus` решена хоть и правильно, но несколько сложновато, можно попроще.

mglukhikh commented:
recheck all

mglukhikh commented:
recheck all

dmitrirazumov committed:
//

dmitrirazumov committed:
//

dmitrirazumov committed:
Program from 6 lesson.

dmitrirazumov committed:
Lesson 7:
  • determining all functions in class and create "function-creators" (task 1)
  • easy program from task 2

dmitrirazumov committed:
Style correction

dmitrirazumov commented:
Программа "PlusMinus" почему-то проходит проверку бота через раз

mglukhikh commented:
Это потому, что бот делает случайные проверки, и вам то попадается тот случай, на котором ваша функция работает неправильно, то нет. По факту это значит, что задача решена с ошибкой

mglukhikh commented on src/lesson7/task1/Matrix.kt:
```diff
@@ -38,32 +41,56 @@ interface Matrix<E> {
+                }
+            }
+        }
+        return true
```
Прошу проверить правильность работы функции `equals` тестами. Тесты добавить в тестовый файл урока 7.1.

mglukhikh replied:
Тестов функции `equals` пока так и не появилось

dmitrirazumov replied:
@mglukhikh Написал тесты, запушил

mglukhikh commented on src/lesson7/task1/Matrix.kt:
```diff
@@ -38,32 +41,56 @@ interface Matrix<E> {
+            }
+        }
+        return true
+    }
```
Отсутствует функция `hashCode()`.

dmitrirazumov replied:
Добавил.

mglukhikh replied:
Плохо добавили. Содержимое списка никак не учитывается.

dmitrirazumov replied:
Надеюсь, что теперь правильно

mglukhikh replied:
Да, так верно

dmitrirazumov commented:
Надо ли мне еще что-то исправлять?

mglukhikh commented on src/lesson5/task1/Parse.kt:
```diff
@@ -118,7 +161,30 @@ fun bestHighJump(jumps: String): Int = TODO()
+        }
+        return result
+    }
+}
```
Функция не всегда проходит тесты. Возьмите один из контр-примеров бота и разберитесь, в чём дело.

mglukhikh commented on src/lesson5/task1/Parse.kt:
```diff
@@ -187,4 +273,25 @@ fun fromRoman(roman: String): Int = TODO()
\ No newline at end of file
+fun computeDeviceCells(cells: Int, commands: String): List<Int> = TODO()
+
+fun myFun(text: String): MutableList<String> {
```
Удалите, пожалуйста, эту функцию из репозитория

mglukhikh commented on src/lesson6/task1/Geometry.kt:
```diff
@@ -55,7 +56,13 @@ data class Circle(val center: Point, val radius: Double) {
-    fun distance(other: Circle): Double = TODO()
+    fun distance(other: Circle): Double {
+        val r_length = (radius + other.radius)
+        val c_length = (center.distance(other.center))
```
Названия `r_length` и `c_length` нарушают code style

mglukhikh commented on src/lesson6/task1/Geometry.kt:
```diff
@@ -70,21 +77,44 @@ data class Circle(val center: Point, val radius: Double) {
+     * Дано множество точек. Вернуть отрезок, соединяющий две наиболее удалённые из них.
+     * Если в множестве менее двух точек, бросить IllegalArgumentException
+     */
+    fun diameter(vararg points: Point): Segment {
```
Сдвиньте эту функцию на 4 пробела влево

mglukhikh commented on src/lesson6/task1/Geometry.kt:
```diff
@@ -70,21 +77,44 @@ data class Circle(val center: Point, val radius: Double) {
+     * Построить окружность по её диаметру, заданному двумя точками
+     * Центр её должен находиться посередине между точками, а радиус составлять половину расстояния между ними
+     */
+    fun circleByDiameter(diameter: Segment): Circle {
```
И эту тоже

mglukhikh commented on test/lesson5/task1/Tests.kt:
```diff
@@ -115,4 +115,9 @@ class Tests {
\ No newline at end of file
+
+    @Test
+    fun myFun() {
```
Тесты `myFun` также прошу удалить из репозитория

mglukhikh commented on src/lesson7/task1/Matrix.kt:
```diff
@@ -38,32 +41,56 @@ interface Matrix<E> {
 
-    override fun toString(): String = TODO()
+    override fun toString(): String {
+        return list.toString()
```
Как будем в этом строковом представлении отличать колонки от рядов?

mglukhikh reviewed (changes requested):
Посмотрел, написал замечаний. Текущая оценка за аттестацию получается 3.5. То есть зачёт у вас уже есть, для получения "3" автоматом нужно пройти тестирование, для получения "4" автоматом нужно также исправить все написанные замечания.

dmitrirazumov committed:
Correction

dmitrirazumov committed:
Add "hashCode()"

dmitrirazumov committed:
Add "hashCode()"

dmitrirazumov commented:
recheck all

dmitrirazumov committed:
Correcting"hashCode()"

dmitrirazumov committed:
Correcting "plusMinus"

dmitrirazumov commented:
Полностью переделал plusMinus, сделал через исключение. Думаю, так правильнее + короче. 
Бота прошел, судя по всему.

dmitrirazumov commented:
Если не секрет, почему вы поставили метку "mediocre"?

mglukhikh commented:
До good (4) осталось проверить `equals` тестами. А пока mediocre (3).

dmitrirazumov committed:
Correcting "equals"

dmitrirazumov committed:
Correcting "equals"

dmitrirazumov committed:
Correcting "equals"

mglukhikh commented on test/lesson7/task1/Tests.kt:
```diff
@@ -49,4 +49,11 @@ class Tests {
+    fun equals() {
+        assertEquals(true, createMatrix(1, 1, listOf(listOf(3))) == createMatrix(1, 1, listOf(listOf(3))))
+        assertEquals(true, createMatrix(2, 2, listOf(listOf(1, 2), listOf(2, 2))) == createMatrix(2, 2, listOf(listOf(1, 2), listOf(2, 2))))
+        assertEquals(false, createMatrix(2, 2, listOf(listOf(1, 1), listOf(2, 2))) != createMatrix(1, 1, listOf(listOf(1, 2), listOf(2, 2))))
```
Этот тест написан некорректно. Вы сравниваете матрицы разных размером, поэтому они должны быть НЕ РАВНЫ, а значит, `false` вначале нужно заменить на `true`. Сделайте это и проверьте, проходит ли тест на вашем компьютере.

dmitrirazumov replied:
@mglukhikh  Исправил

mglukhikh reviewed (changes requested):
Пока вашими тестами я не удовлетворён :(

dmitrirazumov committed:
Correcting "equals"

mglukhikh commented:
Дмитрий, ваша финальная оценка за семестр -- удовлетворительно, с моим мега-сложным заданием про функцию `equals` вы не справились даже со второй попытки. Тест по-прежнему написан неверно. Так что соглашайтесь на "3" автоматом или приходите 16 января на экзамен.

dmitrirazumov commented:
Михаил Игоревич, хотел бы узнать, влияет ли оценка за финальное тестирование на экзаменационную оценку? За сегодняшнее тестирование я получил 5, а за аттестацию (в PR) тройку. 
Дмитрий Хорохорин, 13531/3

mglukhikh commented:
Да, правда, у вас до этого было два теста, написанных на 2. В среднем как раз получается 3.

dmitrirazumov commented:
Когда можно подойти с зачеткой и получить оценку за экзамен?

mglukhikh commented:
То есть вы согласны на "3". Хорошо, я поставлю эту оценку в ведомость. Если вы хотите получить запись в зачётке, то можно подойти на консультацию 13 января (10-00) или на экзамен 16 января (примерно с 10 до 19). Либо ищете одного из преподавателей, ведущих упражнения, они на кафедре чаще меня.
